### PR TITLE
fix: bank deposit behavior

### LINKED ISSUE #1681: bug: single stackable item from inventory to the bank
### Description

Hi, it's my first time for report a bug and I'm not english (I'm sorry for my disaster language) ; 

I found this bug randomly : 

I don't know what's going on, I just want to move my stackable item (potion) to the same item in my bank and some of those same items disappeared suddently
Link (1) : https://youtu.be/knqO2je_GAU

And now, if entierly of my items stacked disapeared in the bank, the slot in my bank is broken and I can't use this slot anymore : 
Link Video (2) :  https://youtu.be/3SLylQVr8y0

### Steps to Reproduce

Step 1 : Drag a stackable item to the bank like a potion or other item (Stackable only) (First Video)
Step 2 : Drag the same item through the inventory into the same slot of the same item in the bank (First Video)
Step 3 : You have the bug and item disapear randomly or/and create randomly too a broken slot in the bank and make it unusable (Second Video)

### Version with bug

v0.7.1.088

### Last version that worked well

Unknow

### Affected platforms

Windows

### Did you find any workaround?

You just shouldn't drag an item from your inventory onto a bank slot with a stackable item, otherwise it would create this bug and probably broke your bank slot. 

### Relevant log output

_No response_

### Duplicate Bug Check

- [X] This bug report is not a duplicate to the best of my knowledge.

### PR DESCRIPTION
**Pull request by Arufonsu**
### Fixes:
- Adjusts quantity and maxQuantity for stackable item's slider.
- Checks if item is stackable when required so we don't unnecessarily show up the deposit slider on single equipment items.
- Context menu bank deposit d/c fix: when destination slot is < 0 (context menu deposit) -> FindsOpenSlot and deposit.
- Non-stackable items fix: we weren't checking stackable items properly, which translated into non-stackable items being placed over each other, leading to glitched bank slots.
- Single stackable items (ie, an arrow) won't overwrite existing item(s) in slot anymore and they stack along with the rest of the same stackable items.

### More fixes along with improvements:
 - We can now easily fill the bank when depositing stackable items, just drag the slider to the right! the server will make sure to adjust the amount and fill the required slots.
 - If the bank slot that an item is dropped on (slot is used as "hint") does not stack or is full of stackable items but there is still available slots in the bank, the server makes sure to find a new slot for non stackable items and fills the non-fully-stacked items.
 - If the bank slot used as "hint" is empty, the server fills the non-fully-stacked items and deposits the rest within the "hint" slot.
 - If there's no initial "hint" (ie, context menu deposit), then finds an open slot to deposit the rest of the stackable items.
 - Updates and simplifies the "CanStore" method to properly work along with this commit changes.
 - Creates the "FindAvailableSpaceForStackableItem" method which finds the amount of available space in the bank for a stackable item.
 - Added in-game system messages to be sent to players when successfully depositing and withdrawing items from the bank.

### Previews

~
(April 15th, 2023)

https://user-images.githubusercontent.com/17498701/232185444-0c5532f0-4cf6-4aa9-80af-fc331f4f9fe6.mp4

~

https://user-images.githubusercontent.com/17498701/231619732-28ca3eca-e3a2-4520-b7b0-a137d926ede7.mp4

~

https://user-images.githubusercontent.com/17498701/231619932-64a53f73-4fc6-4637-8161-4da5d7079e12.mp4

~

https://user-images.githubusercontent.com/17498701/231617217-dd9c7808-d68d-4f12-9525-3600e9001d10.mp4

### CODE REVIEW

**Comment by lodicolo** [comment]
@Arufonsu Both of the videos look like the broken behavior, where is a video of the correct behavior with this fix?

**Comment by lodicolo** [comment]
From Discord:
> If the bank slot does not stack but there is space in the bank still, the move should not fail
> the slot that an item is dropped on should be a "hint" not an absolute "this is where it goes"

That should be addressed too in this PR.

**Comment by Arufonsu** [comment]
Updated PR, code and video previews, so far this manages to fix the issues and behaves as expected, discussed and suggested 👍🏽

**Comment by Arufonsu** [comment]
Updated PR (_again_, sorry for the bunch of force pushes, won't happen after the first review - unless is required).
Fixed some logic issues i've encountered, also added some nice additions, preview:

https://user-images.githubusercontent.com/17498701/232185444-0c5532f0-4cf6-4aa9-80af-fc331f4f9fe6.mp4

Hopefully this yields the expected behavior, PR feels ready for review as well 👍🏽

File: Intersect.Server/Entities/BankInterface.cs [hunk 1]
```diff
@@ -194,55 +194,26 @@ public List<int> FindItemSlots(Guid itemId, int quantity = 1)
         /// </summary>
         /// <param name="item">The <see cref="Item"/> to check for.</param>
         /// <returns>Returns whether or not the bank instance can store this item.</returns>
-        public bool CanStoreItem(Item item)
+        private bool CanStoreItem(Item item)
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:197]: Don't change this to private.

File: Intersect.Server/Entities/BankInterface.cs [hunk 2]
```diff
@@ -194,55 +194,26 @@ public List<int> FindItemSlots(Guid itemId, int quantity = 1)
...
         {
-            if (item.Descriptor != null)
+            // If this item isn't stackable we just need to find an open slot.
+            if (!item.Descriptor.Stackable)
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:200]: Why did you remove the null check...?

File: Intersect.Server/Entities/BankInterface.cs [hunk 3]
```diff
@@ -194,55 +194,26 @@ public List<int> FindItemSlots(Guid itemId, int quantity = 1)
...
+                var bankSlot = mBank[i];
+                if (bankSlot.ItemId == item.ItemId &&
+                    bankSlot.Quantity + item.Quantity <= bankSlot.Descriptor.MaxBankStack)
                 {
-                    if (FindOpenSlots().Count >= item.Quantity)
-                    {
-                        return true;
-                    }
+                    return true;
                 }
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:213]: You removed partial stack filling? If you are trying to put in 100 stackable items, the limit per stack is 100, and you have no totally empty slots but you have two slots that only have 50 which it would fit in, your code rejects it.

**Review by lodicolo — CHANGES_REQUESTED** [review]

File: Intersect.Server/Entities/BankInterface.cs (hunk 3 above)
Comment (Arufonsu) [reply Intersect.Server/Entities/BankInterface.cs:213]: you are right, this happens if there's no slot left that's empty but for example, two slots that aren't completly filled, i've just added the following after the for loop and it seems to handle things well now:

imagen

preview:
https://user-images.githubusercontent.com/17498701/233818612-3a07865c-5033-4316-a31e-2e1f7b47d7dc.mp4

File: Intersect.Server/Entities/BankInterface.cs (hunk 2 above)
Comment (Arufonsu) [reply Intersect.Server/Entities/BankInterface.cs:200]: uuhh ;whoops; ( restored )

File: Intersect.Server/Entities/BankInterface.cs (hunk 3 above)
Comment (Arufonsu) [reply Intersect.Server/Entities/BankInterface.cs:213]: Also: with this, we don't really need the for loop that's before this one condition anymore.

File: Intersect.Server/Entities/BankInterface.cs (hunk 1 above)
Comment (Arufonsu) [reply Intersect.Server/Entities/BankInterface.cs]: restored

**Comment by Arufonsu** [comment]
Self reminder: 
- While playing around with this few weeks ago, i found out a way to also automatically adjust the max quantity sliders in relation with the max stackable items either in bank or in inventory.
- Were we able to drag in many non stackable items with the slider before? if so, i should restore this lol

**Comment by Arufonsu** [comment]
self-reminder: restored the ability to deposit multiple non-stackable items

https://github.com/AscensionGameDev/Intersect-Engine/assets/17498701/f7ab5f5b-f12b-4d52-9231-1d4673736284

PR is ready for review and being built for ease of testing ~

**Comment by Arufonsu** [comment]
Introduces the 'FindAvailableBankSpaceForItem' method to calculate the total available space within the bank for specific items. This enhancement streamlines the bank deposit process by automatically determining the appropriate values for the 'quantity' and 'maxQuantity' parameters in the deposit slider, improving user experience and accuracy when depositing items in the bank.


https://github.com/AscensionGameDev/Intersect-Engine/assets/17498701/a55a39e5-6a4e-4f48-9de2-a943ff874059

**Comment by Arufonsu** [comment]
client enhancement: NoSpace bank system message

- Instead of blindly opening the item deposit slider when there's no space left in the bank, we now inform the player "There is no space left in your bank for that item!" before the slider is up with a max quantity of 0 (which is pointless to do).

https://github.com/AscensionGameDev/Intersect-Engine/assets/17498701/fcae93ee-9a2a-435f-9805-3aaae51dd66a

File: Intersect.Server/Entities/BankInterface.cs [hunk 4]
```diff
@@ -298,6 +271,36 @@ public int FindItemQuantity(Guid itemId)
...
+        /// </summary>
+        /// <param name="itemId">The ID of the item to find available space for.</param>
+        /// <param name="amount">The requested amount of the item to deposit.</param>
+        /// <returns>The maximum amount of the item that can be deposited in the bank.</returns>
+        private int FindAvailableSpaceForItem(Guid itemId, int amount)
+        {
+            int spaceLeft = 0;
+            for (int i = 0; i < mMaxSlots; i++)
+            {
+                if (mBank[i]?.ItemId == itemId)
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:290]: Before the if statement add this as a short-circuit to skip costly searches

```
if (spaceLeft >= amount)
{
    return amount;
}
```

File: Intersect.Server/Entities/BankInterface.cs [hunk 5]
```diff
@@ -196,53 +196,26 @@ public List<int> FindItemSlots(Guid itemId, int quantity = 1)
...
-                        if (Math.Ceiling((double)item.Quantity / item.Descriptor.MaxBankStack) <= FindOpenSlots().Count)
-                        {
-                            return true;
-                        }
-                    }
-                }
-                // Not a stacking item, so can we contain the amount we want to give them?
-                else
-                {
-                    if (FindOpenSlots().Count >= item.Quantity)
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:237]: This case appears to not be covered in the new code?

File: Intersect.Server/Entities/BankInterface.cs [hunk 6]
```diff
@@ -319,79 +322,106 @@ public bool TryDepositItem(int slot, int amount, int destSlot = -1, bool sendUpd
...
+                PacketSender.SendChatMsg(mPlayer, Strings.Items.nobank, ChatMessageType.Bank, CustomColors.Items.Bound);
+                return false;
+            }
 
-                        if (CanStoreItem(new Item(itemBase.Id, amount), destSlot))
+            // Calculate the maximum amount that can be deposited, based on quantity and bank space.
+            var depositMultipleItems = mPlayer.FindInventoryItemQuantity(inventoryItemItemId) > 1;
+            var maxAmount = depositMultipleItems ? Math.Min(mPlayer.FindInventoryItemQuantity(inventoryItemItemId),
+                    FindAvailableSpaceForItem(inventoryItemItemId, amount)) : 1;
+            amount = Math.Min(amount, maxAmount);
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:345]: `depositMultipleItems` should be based on this line instead?

File: Intersect.Server/Entities/BankInterface.cs [hunk 7]
```diff
@@ -319,79 +322,106 @@ public bool TryDepositItem(int slot, int amount, int destSlot = -1, bool sendUpd
...
+            {
+                PacketSender.SendChatMsg(mPlayer, Strings.Items.nobank, ChatMessageType.Bank, CustomColors.Items.Bound);
+                return false;
+            }
 
-                        if (CanStoreItem(new Item(itemBase.Id, amount), destSlot))
+            // Calculate the maximum amount that can be deposited, based on quantity and bank space.
+            var depositMultipleItems = mPlayer.FindInventoryItemQuantity(inventoryItemItemId) > 1;
+            var maxAmount = depositMultipleItems ? Math.Min(mPlayer.FindInventoryItemQuantity(inventoryItemItemId),
+                    FindAvailableSpaceForItem(inventoryItemItemId, amount)) : 1;
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:344]: `mPlayer.FindInventoryItemQuantity(inventoryItemItemId)` is repeated in both the condition and the branch, the branch is totally unnecessary here.

Comment (lodicolo) [reply Intersect.Server/Entities/BankInterface.cs:344]: This could be changed to
```csharp
var inventoryQuantity = mPlayer.FindInventoryItemQuantity(inventoryItemItemId);
amount = Math.Min(amount, inventoryQuantity); // This is to optimize the next line
var availableStorageSpace = FindAvailableSpaceForItem(inventoryItemItemId, amount));
var maxAmount = Math.Min(inventoryQuantity, availableStorageSpace);
```

File: Intersect.Server/Entities/BankInterface.cs [hunk 8]
```diff
@@ -319,79 +322,106 @@ public bool TryDepositItem(int slot, int amount, int destSlot = -1, bool sendUpd
...
+                    destSlot = FindOpenSlot();
+                }
+
+                if (destSlot < 0)
+                {
+                    PacketSender.SendChatMsg(mPlayer, Strings.Banks.banknospace, ChatMessageType.Bank,
+                        CustomColors.Alerts.Error);
+                    return false;
                 }
             }
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:385]: I'm wondering if generic and simpler logic is better than introducing an "if 1" condition

File: Intersect.Server/Entities/BankInterface.cs [hunk 9]
```diff
@@ -319,79 +322,106 @@ public bool TryDepositItem(int slot, int amount, int destSlot = -1, bool sendUpd
...
+            }
 
-                        if (CanStoreItem(new Item(itemBase.Id, amount), destSlot))
+            // Calculate the maximum amount that can be deposited, based on quantity and bank space.
+            var depositMultipleItems = mPlayer.FindInventoryItemQuantity(inventoryItemItemId) > 1;
+            var maxAmount = depositMultipleItems ? Math.Min(mPlayer.FindInventoryItemQuantity(inventoryItemItemId),
+                    FindAvailableSpaceForItem(inventoryItemItemId, amount)) : 1;
+            amount = Math.Min(amount, maxAmount);
+
+            if (amount <= 0)
```
Comment (lodicolo) [inline Intersect.Server/Entities/BankInterface.cs:347]: amount < 1

**Review by lodicolo — CHANGES_REQUESTED** [review]
This isn't an exhaustive review but there is a missing thing and a few over-complicated things that I think can be simplified

File: Intersect.Server/Entities/BankInterface.cs (hunk 5 above)
Comment (Arufonsu) [reply Intersect.Server/Entities/BankInterface.cs:237]: This should be covering it when it comes to non stackable items:

https://github.com/AscensionGameDev/Intersect-Engine/blob/0765628603c5e32890fc7757e6dc79f2394fa100/Intersect.Server/Entities/BankInterface.cs#L205-L209

File: Intersect.Server/Entities/BankInterface.cs (hunk 9 above)
Comment (Arufonsu) [reply Intersect.Server/Entities/BankInterface.cs:347]: maxAmount < 1 after the suggested changes (?)

Comment (lodicolo) [reply Intersect.Server/Entities/BankInterface.cs:347]: No, it would still be `amount < 1` because you would not be getting rid of the last `amount = Math.Min(amount, maxAmount);`

File: Intersect.Server/Entities/BankInterface.cs (hunk 5 above)
Comment (lodicolo) [reply Intersect.Server/Entities/BankInterface.cs:237]: Yeah but the code specifically said `>= item.Quantity` not `1` -- it makes me think there's a case where something can be not stackable with a quantity > 1...

Comment (Arufonsu) [reply Intersect.Server/Entities/BankInterface.cs:237]: that would be weird (?) but just to make sure, i can add this check back, although i would have to see how it behaves with this... which tbh feels weird, right now we are just looking away for an available slot whenever it comes to stackables, since they aren't supposed to ... stack together in a single a slot (?)
